library/policy_enforcer.js

```diff
@@ -0,0 +1,258 @@
+      try {
+        await member.setMute(false)
+      } catch (err) {
+        // maybe they left already.
+        this.logFn_(err)
+      }
+    }
+  }
+
+  // auto-VC creation: create a room if all rooms are occupied. Muted or unmuted doesn't matter,
+  // because in general we want to discourage people from using rooms that are occupied even if all
+  // the participants are currently muted. Remove a temp room if there are at least two empty ones.
+  async createRemoveRooms (guild) {
+    let emptyRooms =
+      guild.channels.filter(chan => chan.isPermittedChannel && !chan.members.some(m => !m.deleted))
+
+    if (emptyRooms.size === 0) {
+      // no empty rooms; create a new channel
+      const categoryChan = findChannel(guild, 'practice-room-chat').parent
+      const pinanoBot = findRole(guild, 'Pinano Bot')
+      const tempMuted = findRole(guild, 'Temp Muted')
+      const verifRequired = findRole(guild, 'Verification Required')
+      const everyone = findRole(guild, '@everyone')
+
+      let newChan = await guild.createChannel('Practice Room', {
+        type: 'voice',
+        parent: categoryChan,
+        bitrate: settings.dev_mode ? 96000 : 384000,
+        position: categoryChan.children.size,
+        permissionOverwrites: [{
+          id: pinanoBot,
+          allow: ['MANAGE_CHANNELS', 'MANAGE_ROLES']
+        }, {
+          id: tempMuted,
+          deny: ['SPEAK']
+        }, {
+          id: verifRequired,
+          deny: ['VIEW_CHANNEL']
+        }, {
+          id: everyone,
+          deny: ['MANAGE_CHANNELS', 'MANAGE_ROLES']
+        }]
+      })
+
+      newChan.isPermittedChannel = true
+      newChan.isTempRoom = true
+
+      // update the db
+      await this.guildRepository_.addToField(guild.id, 'permitted_channels', newChan.id)
+    } else if (emptyRooms.size >= 2) {
+      // remove an extra room if 1) there are at least two empty rooms and 2) one of those rooms is
+      // a temp room. (We don't want to destroy the primary rooms.)
+      let emptyRoom = emptyRooms.find(c => c.name === 'Practice Room' || c.isTempRoom)
+      if (emptyRoom != null) {
+        // before removing the channel from the guild, remove it in the db.
+        await this.guildRepository_.removeFromField(guild.id, 'permitted_channels', emptyRoom.id)
+        await emptyRoom.delete()
+        return emptyRoom
+      }
+    }
+  }
+
+  // members in any practice channel who are not deaf-muted may speak in #practice-room-chat.
+  async applyPermissions (guild, member) {
+    let prChat = findChannel(guild, 'practice-room-chat')
+    if (prChat == null) {
+      return
+    }
+
+    if (member.voiceChannel != null &&
+      member.voiceChannel.isPermittedChannel &&
+      !(member.mute && member.selfDeaf) &&
+      !isTempMuted(member)) {
+      await prChat.overwritePermissions(member.id, { SEND_MESSAGES: true })
+    } else {
+      prChat = await prChat.overwritePermissions(member.id, { SEND_MESSAGES: null })
```

Comment:
discord.js has a bug here (#3570). The result of `overwritePermissions` may or may not be reflected in the return value.